lib/eig_block_trlm.cpp

```diff
@@ -0,0 +1,771 @@
+      }
+      saveVectors(vecs_ptr, eig_param->vec_outfile);
+    }
+
+    if (getVerbosity() >= QUDA_SUMMARIZE) {
+      printfQuda("***********************************\n");
+      printfQuda("***** END BLOCK TRLM SOLUTION *****\n");
+      printfQuda("***********************************\n");
+    }
+
+    // Save BLOCK TRLM tuning
+    saveTuneCache();
+
+    mat.flops();
+  }
+
+  // Destructor
+  BLKTRLM::~BLKTRLM()
+  {
+    host_free(jth_block);
+    host_free(block_alpha);
+    host_free(block_beta);
+  }
+
+  // Block Thick Restart Member functions
+  //---------------------------------------------------------------------------
+  void BLKTRLM::blockLanczosStep(std::vector<ColorSpinorField *> v, int j)
+  {
+    // Compute r = A * v_j - b_{j-i} * v_{j-1}
+
+    // Offset for alpha, beta matrices
+    int arrow_offset = j * block_size;
+    int idx = 0, idx_conj = 0;
+
+    // r = A * v_j
+    for (int b = 0; b < block_size; b++) chebyOp(mat, *r[b], *v[j + b]);
+
+    // r = r - b_{j-1} * v_{j-1}
+    int start = (j > num_keep) ? j - block_size : 0;
+    if (j - start > 0) {
+
+      std::vector<ColorSpinorField *> r_;
+      r_.reserve(block_size);
+      for (int i = 0; i < block_size; i++) r_.push_back(r[i]);
+
+      int blocks = (j - start) / block_size;
+      std::vector<Complex> beta_;
+      beta_.reserve(blocks * block_data_length);
+
+      // Switch beta block order from COLUMN to ROW major
+      // This switches the block from upper to lower triangular
+      for (int i = 0; i < blocks; i++) {
+        int block_offset = (i + start / block_size) * block_data_length;
+        for (int b = 0; b < block_size; b++) {
+          for (int c = 0; c < block_size; c++) {
+            idx = c * block_size + b;
+            beta_.push_back(-block_beta[block_offset + idx]);
+          }
+        }
+      }
+
+      std::vector<ColorSpinorField *> v_;
+      v_.reserve(j - start);
+      for (int i = start; i < j; i++) { v_.push_back(v[i]); }
+      if (blocks == 1)
+        blas::caxpy_L(beta_.data(), v_, r_);
+      else
+        blas::caxpy(beta_.data(), v_, r_);
+    }
+
+    // a_j = v_j^dag * r
+    std::vector<ColorSpinorField *> vecs_ptr;
+    vecs_ptr.reserve(block_size);
+    for (int b = 0; b < block_size; b++) { vecs_ptr.push_back(v[j + b]); }
+    // Block dot products stored in alpha_block.
+    blas::cDotProduct(block_alpha + arrow_offset, vecs_ptr, r);
+
+    // Use jth_block to negate alpha data and apply block BLAS.
+    // Data is in square hermitian form, no need to switch to ROW major
+    for (int b = 0; b < block_size; b++) {
+      for (int c = 0; c < block_size; c++) {
+        idx = b * block_size + c;
+        jth_block[idx] = -1.0 * block_alpha[arrow_offset + idx];
+      }
+    }
+
+    // r = r - a_j * v_j
+    blas::caxpy(jth_block, vecs_ptr, r);
+
+    // Orthogonalise R[0:block_size] against the Krylov space V[0:j + block_size]
+    for (int k = 0; k < 1; k++) blockOrthogonalize(v, r, j + block_size);
```

Comment:
Is this loop here for potential future exploration with multiple passes of orthogonalization?

Reply:
I think, during the time I was exploring Implicitly Restarted Lanczos, I had to crank this up to maintain orthonormality. TRLM is very stable though, so this can go.

Reply:
You could always expose the number of orthogonalization sweeps as a parameter.  We've done in this multigrid.  Might be nice to leave this in there and just default the value to 1.